query-builder/pom.xml

```diff
@@ -0,0 +1,84 @@
+<!--
+
+    Copyright DataStax, Inc.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>com.datastax.oss</groupId>
+    <artifactId>java-driver-parent</artifactId>
+    <version>4.0.0-alpha4-SNAPSHOT</version>
+  </parent>
+
+  <artifactId>java-driver-query-builder</artifactId>
+  <packaging>jar</packaging>
+
+  <name>DataStax Java driver for Apache Cassandra® - query builder</name>
+
+  <dependencies>
+    <dependency>
+      <groupId>com.datastax.oss</groupId>
+      <artifactId>java-driver-core</artifactId>
+      <version>${project.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>com.google.guava</groupId>
+      <artifactId>guava</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.assertj</groupId>
+      <artifactId>assertj-core</artifactId>
+      <scope>test</scope>
+    </dependency>
+  </dependencies>
+
+  <build>
+    <plugins>
+      <plugin>
+        <artifactId>maven-shade-plugin</artifactId>
```

Comment:
I think we need a better solution for shading guava.  I.E. if you are using the query builder, you will now have two instances of shaded guava packages in your application.

Would it makes sense to have a separate artifact that only includes the shaded guava packages? This way we don't need to shade in every module we want to use guava in.

Reply:
Tinkerpop does something like that, but that means you need to re-install the shaded artifact manually in your local repository every time you bump the version, otherwise `mvn compile` in individual modules doesn't work. I don't like it, anything that disrupts the standard Maven lifecycle is bad practice IMO.

Is it really a problem to re-shade Guava in every module? It's "only" an extra 2.6MB.

One thing we could do is exclude the classes that are not used at runtime (à la ProGuard), but that seems a bit dangerous.

Reply:
> create an artifact com.datastax.oss:guava-shaded:21.0

Good idea 👍

Reply:
Looking around a bit and I noticed that Apache Flink does this for a number of dependencies for this very reason: https://github.com/apache/flink-shaded

> This repository contains a number of shaded dependencies for the Apache Flink project.
>
> The purpose of these dependencies is to provide a single instance of a shaded dependency in the Flink distribution, instead of each individual module shading the dependency.

Other than that I haven't found another prominent project doing the same, but I think this could the best solution if we want to use guava.

Reply:
Yes, I'm 👍 on a separate project. It only needs Guava at this stage, but we could add other libraries in the future if we also need to shade them.